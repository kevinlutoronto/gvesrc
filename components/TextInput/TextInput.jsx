// @flow
import React, { Component } from 'react';
import cx from 'classnames';

import { getDataAttrs } from '../../dataUtils';
import type {
  ChangeEventHandlers,
  Data,
  FocusEventHandlers,
  GlobalAttributes,
  SyntheticMouseEventHandler,
} from '../../types';

import styles from './style.scss';

export type Props = {|
  data?: Data,
  disabled?: boolean,
  error?: boolean,
  maxLength?: number | string,
  name?: string,
  placeholder?: string,
  prefix?: React$Node,
  suffix?: React$Node,
  type?: 'text' | 'password' | 'email' | 'url',
  value?: string,
  onClick?: SyntheticMouseEventHandler<HTMLInputElement>,
  ...FocusEventHandlers<HTMLInputElement>,
  ...ChangeEventHandlers<HTMLInputElement>,
  ...GlobalAttributes,
|};

class TextInput extends Component<Props> {
  static defaultProps = {
    placeholder: 'Type something...',
  };

  render() {
    const { className, disabled, error, data, prefix, suffix, type, ...rest } = this.props;

    const inputType = type || 'text';

    const classes = cx(styles.wrapper, className, {
      [styles.wrapperDisabled]: disabled,
      [styles.wrapperError]: error,
    });

    return (
      <div {...getDataAttrs(data)} className={classes}>
        {prefix && <div className={styles.prefix}>{prefix}</div>}
        <input className={styles.input} disabled={disabled} type={inputType} {...rest} />
        {suffix && <div className={styles.suffix}>{suffix}</div>}
      </div>
    );
  }
}

TextInput.displayName = 'Plasma@TextInput';

export default TextInput;
