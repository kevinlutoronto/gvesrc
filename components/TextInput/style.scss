@import '../../styles/base';

.wrapper {
  @extend %input--wrapper;
  @extend %type--body;

  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  font-size: $typeBase;
  justify-content: stretch;

  &:hover {
    @extend %input--wrapper--hover;

    .prefix,
    .suffix {
      border-color: $cGray200;
    }
  }

  &:focus-within {
    @extend %input--wrapper--focus;

    .prefix,
    .suffix {
      border-color: $cGray200;
    }
  }

  &.wrapperDisabled {
    @extend %input--wrapper--disabled;

    .prefix,
    .suffix {
      border-color: $cGray400;
    }
  }

  &.wrapperError,
  &.wrapperError:hover {
    @extend %input--wrapper--error;

    .prefix,
    .suffix {
      border-color: $cWarningRed;
    }
  }
}

.input {
  @extend %input;
  @extend %type--body;

  border: 0;
  flex: 1 1 auto;
  outline: none;

  &:disabled {
    @extend %input--disabled;
  }
}

.prefix,
.suffix {
  border: 0 solid $cGray300;
  background-color: $cGray600;
  flex: 0 0 auto;
  padding: 10px;
}

.suffix {
  border-left-width: 1px;
  border-radius: 0 $inputRadius $inputRadius 0;
}

.prefix {
  border-right-width: 1px;
  border-radius: $inputRadius 0 0 $inputRadius;
}
