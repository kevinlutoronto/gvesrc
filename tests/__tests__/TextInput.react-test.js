/* eslint-disable */

import React from 'react';
import TextInput from '../../components/TextInput/TextInput';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

describe('TextInput Component', () => {
  test('render', () => {
    const component = renderer.create(
      <TextInput name="name" id="id" value="foo" maxLength={100} />
    );
    expect(component).toMatchSnapshot();
  });

  test('render placeholder', () => {
    const component = renderer.create(<TextInput placeholder="placeholder" />);
    expect(component).toMatchSnapshot();
  });

  test('render prefix', () => {
    const component = renderer.create(<TextInput prefix="prefix" />);
    expect(component).toMatchSnapshot();
  });

  test('render suffix', () => {
    const component = renderer.create(<TextInput suffix="suffix" />);
    expect(component).toMatchSnapshot();
  });

  test('render error', () => {
    const component = renderer.create(<TextInput error />);
    expect(component).toMatchSnapshot();
  });

  test('change', () => {
    const onChange = jest.fn();
    const component = mount(<TextInput onChange={onChange} />);

    const input = component.find('input');

    const value = 'test';

    input.getDOMNode().value = value;
    input.simulate('change');

    expect(onChange).toHaveBeenCalledTimes(1);

    expect(onChange.mock.calls[0][0]).toBeDefined();
    expect(onChange.mock.calls[0][0].target.value).toBe(value);
  });

  test('focus and blur', () => {
    const onFocus = jest.fn();
    const onBlur = jest.fn();

    const component = mount(<TextInput onFocus={onFocus} onBlur={onBlur} />);

    const input = component.find('input');

    input.simulate('focus');
    expect(onFocus).toHaveBeenCalledTimes(1);

    input.simulate('blur');
    expect(onBlur).toHaveBeenCalledTimes(1);
  });
});
