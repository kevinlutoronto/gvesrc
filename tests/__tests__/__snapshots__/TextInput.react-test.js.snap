// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component render 1`] = `
<div
  className="wrapper"
>
  <input
    className="input"
    disabled={undefined}
    id="id"
    maxLength={100}
    name="name"
    placeholder="Type something..."
    type="text"
    value="foo"
  />
</div>
`;

exports[`TextInput Component render error 1`] = `
<div
  className="wrapper wrapperError"
>
  <input
    className="input"
    disabled={undefined}
    placeholder="Type something..."
    type="text"
  />
</div>
`;

exports[`TextInput Component render placeholder 1`] = `
<div
  className="wrapper"
>
  <input
    className="input"
    disabled={undefined}
    placeholder="placeholder"
    type="text"
  />
</div>
`;

exports[`TextInput Component render prefix 1`] = `
<div
  className="wrapper"
>
  <div
    className="prefix"
  >
    prefix
  </div>
  <input
    className="input"
    disabled={undefined}
    placeholder="Type something..."
    type="text"
  />
</div>
`;

exports[`TextInput Component render suffix 1`] = `
<div
  className="wrapper"
>
  <input
    className="input"
    disabled={undefined}
    placeholder="Type something..."
    type="text"
  />
  <div
    className="suffix"
  >
    suffix
  </div>
</div>
`;
